tests/thread_test.c

```diff
@@ -85,3 +86,61 @@ static int s_test_thread_atexit(struct aws_allocator *allocator, void *ctx) {
 }
 
 AWS_TEST_CASE(thread_atexit_test, s_test_thread_atexit)
+
+struct managed_thread_test_data {
+    uint64_t sleep_time_in_ns;
+};
+
+static void s_managed_thread_fn(void *arg) {
+    struct thread_test_data *test_data = (struct thread_test_data *)arg;
+    test_data->thread_id = aws_thread_current_thread_id();
+}
+
+#define MAX_MANAGED_THREAD_TEST_QUANTITY 16
+
+static int s_do_managed_thread_join_test(struct aws_allocator *allocator, size_t thread_count) {
+
+    struct aws_thread threads[MAX_MANAGED_THREAD_TEST_QUANTITY];
+    struct managed_thread_test_data thread_data[MAX_MANAGED_THREAD_TEST_QUANTITY];
+
+    AWS_FATAL_ASSERT(thread_count <= MAX_MANAGED_THREAD_TEST_QUANTITY);
+
+    for (size_t i = 0; i < thread_count; ++i) {
+        thread_data[i].sleep_time_in_ns =
+            aws_timestamp_convert(100 * (i / 2), AWS_TIMESTAMP_MILLIS, AWS_TIMESTAMP_NANOS, NULL);
+        aws_thread_init(&threads[i], allocator);
+    }
+
+    struct aws_thread_options thread_options = *aws_default_thread_options();
+    thread_options.join_strategy = AWS_TJS_MANAGED;
+
+    for (size_t i = 0; i < thread_count; ++i) {
+        ASSERT_SUCCESS(
+            aws_thread_launch(&threads[i], s_managed_thread_fn, (void *)&thread_data[i], &thread_options),
+            "thread creation failed");
+
+        aws_thread_clean_up(&threads[i]);
+
+        ASSERT_INT_EQUALS(
+            AWS_THREAD_MANAGED, aws_thread_get_detach_state(&threads[i]), "thread state should have returned JOINABLE");
+    }
+
+    aws_thread_join_all_managed();
```

Comment:
This MUST be called from aws_common_library_clean_up() and it should be idempotent. If user code wants to call it, that's fine, but cleaning up the library should ENFORCE it. This test, for instance, should not have to call it directly.

Reply:
I'm a little uneasy making this mandatory within cleanup.  The join is unbounded and if the user makes a mistake with a thread it's an infinite wait.

Reply:
I haven't reviewed the changes, but this comment just caught my eye. I don't like the idea of the SDK being unsafe by default and it seemed that on [Oct 9th](https://github.com/awslabs/aws-crt-cpp/issues/190#issuecomment-706233116) in  awslabs/aws-crt-cpp#190 you agreed:
>First possibility that jumps to mind is a possible flag in the API object that controls blocking or non-blocking shutdown with a safe (blocking) default.

If the flag is set to unblocking, it would be also nice to have a timeout in the `aws_thread_join_all_managed()` call so that the caller can wait for a bounded time and then bail out in application specific manner, if needed.

Reply:
Over time, I've found myself going the other way.

An infinite wait is the same as a deadlock and represents a fundamental programming mistake somewhere, either ours or the users.

Waiting to a timeout is just as untrustworthy as not fixing the crash in that if you hit the condition you'll likely crash afterwards when you unload the dll, and if waiting for a timeout was preferable (or fixes the problem, after all we're talking about millisecond or lower level timing issues relative to the dll unload) that could have been done before this change.

We wait in aws_common_library_clean_up now and so if you absolutely cannot do the unbounded wait there'd need to be a new api to control that duration.  At this stage I'd prefer to leave this as is but if you guys feel strongly about the need I'll see if there's any objections to adding it.

Reply:
This is the safe by default choice. We shut everything we own down before returning. There's no safe way to shutdown with threads still out there accessing shared state, which we are about to clean up synchronously. If there's a deadlock or fundamental programming error, then that is a bug, that's what we have the test suite for. We are going for reliable correctness here, not "how do we let the user do something unsafe when resources have not been cleaned up".

Reply:
Normally I would say wait and see, but given the difficulty/rarity of deploying to a device, any feature added after release may as well not exist.